[GOL-2] feat(checkbox): Implement Checkbox component according to Material Design 3

JLChambiM opened the pull request:
# Checkbox Component Implementation (GOL-2)

## Description
This PR implements a reusable Checkbox component following Material Design 3 specifications as required in ticket GOL-2.

## Changes
- Created the Checkbox component with proper states (checked/unchecked/disabled)
- Implemented Material Design 3 styles
- Added Storybook configuration for component visualization
- Implemented comprehensive test coverage



## Testing
All tests are passing, verifying that the component:
- Renders correctly when not checked
- Renders correctly when checked
- Calls onChange with the new state when clicked

## References
- GOL-2
- [Material Design 3 Checkbox](https://m3.material.io/components/checkbox/overview)

Commit:
feat(checkbox): implement Checkbox component according to Material Design 3

- Created Checkbox component following Material Design 3 specifications
- Implemented all required states: checked, unchecked, disabled
- Added Storybook configuration for component visualization
- Added full test coverage verifying:
  - Renders correctly when not checked
  - Renders correctly when checked
  - Calls onChange with the new state when clicked

References: GOL-2

ougarte commented:
@JLChambiM  I recommend renaming the branch to `feature/GOL-2`

ougarte commented on apps/ui/src/components/checkbox/__tests__/checkbox.test.tsx:
```diff
@@ -0,0 +1,28 @@
+import '@testing-library/jest-dom';
+import { Checkbox } from '../checkbox';
+
+describe('Checkbox', () => {
```
I recommend renaming from `Checkbox ` to `components/checkbox`

ougarte commented on apps/ui/src/components/checkbox/__tests__/checkbox.test.tsx:
```diff
@@ -0,0 +1,28 @@
+import { Checkbox } from '../checkbox';
+
+describe('Checkbox', () => {
+	it('renders correctly when not checked', () => {
```
I recommend moving `it` into a describe as follow:

```tsx
describe('components/checkbox', () => {
    describe('layout', () => {
        test('should render by default', () => {
            // arrange
            render(<Checkbox checked={false} onChange={() => { /* No-op */ }} />);

            // act
            const checkbox = screen.getByRole('checkbox');

            // assert
            expect(checkbox).not.toBeChecked();
        });
        test('should render as checked', () => {
            // arrange
            // act
            // assert
        });
    });

    describe('behavior', () => {
         test(`should trigger  'onChange' event`, () => {
            // arrange
            // act
            // assert
         });
    });
})
```

ougarte commented on apps/ui/src/components/checkbox/helpers.tsx:
```diff
@@ -0,0 +1,3 @@
+export const generateId = (): string => {
+	return `checkbox-${Math.random().toString(36).slice(2, 9)}`;
+};
```
I recommend adding a `toDefaults` function as follow

```ts
export function toDefaults(properties?: Properties): Required<Properties> {
    return _.defaults({}, properties, {
        id: generateId(),
        checked: false,
        disabled: false,
        onChange: _.noop,
    }) as Required<Properties>;
}
```

ougarte commented on apps/ui/src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,48 @@
+	};
+
+	return (
+		<div style={{ display: 'flex', alignItems: 'center' }}>
```
I recommend using `classNames` instead of `style`

ougarte commented on apps/ui/src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,48 @@
+
+	return (
+		<div style={{ display: 'flex', alignItems: 'center' }}>
+			<div className="checkbox-container">
```
I recommend using `container` instead of `checkbox-container`

ougarte commented on apps/ui/src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,48 @@
+				/>
+			</div>
+			{label && (
+				<label className="checkbox-label" htmlFor={checkboxId}>
```
I recommend using `label` instead of `checkbox-label`

ougarte commented on apps/ui/src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,48 @@
+	disabled = false,
+	onChange,
+}) => {
+	const [isChecked, setIsChecked] = useState(checked);
```
I do not recommend using `useState` instead use `useRef`

ougarte commented on apps/ui/src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,48 @@
+import './styles/checkbox.css';
+import { generateId } from './helpers';
+
+export const Checkbox: React.FC<CheckboxProperties> = ({ 
```
I recommend using function approach as follow

```tsx
export default function Checkbox(properties?: Properties) {
     const defaults = toDefaults(properties);
     // ...code
}
```

ougarte commented on apps/ui/src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,48 @@
+		setIsChecked(checked);
+	}, [checked]);
+
+	const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
```
I recommend renaming from `handleChange` to `onChangeHandler`

```tsx
const onChangeHandler = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
    defaults.onChange(event, event.target.checked);
}, []);
```

ougarte commented on apps/ui/src/components/checkbox/types.ts:
```diff
@@ -0,0 +1,7 @@
+export interface CheckboxProperties {
```
I recommend using `type` instead of `interface` as follow
```ts
export type Properties = {
  id?: string;
  label?: string;
  checked?: boolean;
  disabled?: boolean;
  onChange: (checked: boolean) => void;
}
```

ougarte commented on apps/ui/src/components/checkbox/styles/checkbox.css:
```diff
@@ -0,0 +1,98 @@
+.checkbox-container {
```
I recommend moving all styles into `checkbox` scope as follow

```css
.checkbox {
    .container {}

    .input {
       &:checked {}
       &:checked::after {}
    }
    .label {}

    /* ... more styles */
}
```

ougarte commented on apps/ui/src/components/checkbox/styles/checkbox.css:
```diff
@@ -0,0 +1,98 @@
+  align-items: center;
+  justify-content: center;
+  width: 48px; /* Target size */
+  height: 48px;
```
I recommend using css variables ex: `--checkbox-size`

ougarte commented on apps/ui/src/components/checkbox/styles/checkbox.css:
```diff
@@ -0,0 +1,98 @@
+  appearance: none;
+  position: absolute;
+  width: 18px; /* Container size */
+  height: 18px;
```
I recommend using **css variables**

Commit:
refactor: update checkbox component with improved implementation

ougarte commented on apps/ui/src/components/checkbox/__tests__/checkbox.test.tsx:
```diff
@@ -0,0 +1,67 @@
+import '@testing-library/jest-dom';
+import Checkbox from '../checkbox';
+
+describe('Checkbox/checkbox', () => {
```
I recommend updating as `components/checkbox`

ougarte commented on apps/ui/src/components/checkbox/__tests__/checkbox.test.tsx:
```diff
@@ -0,0 +1,67 @@
+describe('Checkbox/checkbox', () => {
+	describe('layout', () => {
+		test('should render by default', () => {
+			render(<Checkbox checked={false} onChange={() => {/* No-op */}} />);
```
I recommend removing `checked, onChange` properties so that, the properties will be set as defaults

ougarte commented on apps/ui/src/components/checkbox/__tests__/checkbox.test.tsx:
```diff
@@ -0,0 +1,67 @@
+			expect(checkbox).not.toBeChecked();
+		});
+		test('should render as checked', () => {
+			render(<Checkbox checked={true} onChange={() => {/* No-op */}} />);
```
I recommend removing `onChange` property, so that it should be set as defualt

ougarte commented on apps/ui/src/components/checkbox/__tests__/checkbox.test.tsx:
```diff
@@ -0,0 +1,67 @@
+			const checkbox = screen.getByRole('checkbox');
+			expect(checkbox).toBeChecked();
+		});
+		test('should render with label when provided', () => {
```
I recommend updating as `should display label property`

ougarte commented on apps/ui/src/components/checkbox/__tests__/checkbox.test.tsx:
```diff
@@ -0,0 +1,67 @@
+			const label = screen.getByText('Test Label');
+			expect(label).toBeInTheDocument();
+		});
+		test('should apply disabled styles when disabled', () => {
```
I recommend updating as `should display as disabled`

ougarte commented on apps/ui/src/components/checkbox/__tests__/checkbox.test.tsx:
```diff
@@ -0,0 +1,67 @@
+			
+			expect(onChange).toHaveBeenCalled();
+		});
+		test('should trigger onChange event when container is clicked', () => {
```
[question] what does `container` mean ?? - _please add more details about it..._

ougarte commented on apps/ui/src/components/checkbox/__tests__/checkbox.test.tsx:
```diff
@@ -0,0 +1,67 @@
+			
+			expect(onChange).toHaveBeenCalled();
+		});
+		test('should not trigger onChange when disabled', () => {
```
I recommend updating as `should not trigger onChange event when it is disabled`

ougarte commented on apps/ui/src/components/checkbox/styles/base.css:
```diff
@@ -0,0 +1,26 @@
+  /* input reset */
+  appearance: none;
+  -webkit-appearance: none;
+  -moz-appearance: none;
```
I recommend removing `vendor prefixies`  because the PostCSS - autoprefixes added them automatically.

ex(`-webkit-*, -moz-*`)

ougarte commented on apps/ui/src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,63 @@
+	}, [defaults.disabled]);
+
+	return (
+		<div className={getContainerClass(defaults.theme, defaults.disabled)}>
```
I recommend adding `checkbox` class by default on primary element

ougarte commented on apps/ui/src/components/checkbox/styles/base.css:
I recommend moving all styles under a primary class (ex: '.checbox')

ougarte commented on apps/ui/src/components/checkbox/styles/layout.css:
I recommend moving all styles under `.checkbox` namespace

ougarte commented on apps/ui/src/components/checkbox/styles/state.css:
I recommend moving all styles under `.checkbox` namespace

ougarte commented on apps/ui/src/components/checkbox/styles/theme.css:
I think we need to review that `theme` means...

ougarte commented on apps/ui/src/components/checkbox/types.ts:
```diff
@@ -0,0 +1,34 @@
+     * 
+     * @default 'primary'
+     */
+    theme?: 'primary' | 'secondary' | 'error';
```
I recommend using `variant` property instead of `theme`

Commit:
feat(checkbox): add indeterminate state

Applied recommendations and implemented indeterminate checkbox state
with accessibility and variant styling.

ougarte commented on apps/ui/src/components/checkbox/styles/variables.css:
```diff
@@ -0,0 +1,36 @@
+:root {
```
I recommend using a prefix called `--checkbox-**` on **all checkbox css  variables** ex: `--checkbox-color-primary` so that it wont be a conflict with other css variables

ougarte commented on apps/ui/src/components/checkbox/styles/variant.css:
```diff
@@ -0,0 +1,68 @@
+/* variant: Primary */
+.checkbox.variant-primary .input:checked {
```
I recommend removing the `variant-` prefix on your styles

ougarte commented on apps/ui/src/components/checkbox/types.ts:
```diff
@@ -0,0 +1,40 @@
+  /**
+   *
+   */
+  onChange?: (checked: boolean) => void;
```
I recommend changing as follow

```tsx
onChange?: (event: React.StaticEvent, properties: Properties) => void;
```

ougarte commented on apps/ui/src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,69 @@
+
+	const inputReference = useRef<HTMLInputElement>(null);
+
+	useEffect(() => {
```
I recommend using `useLayoutEffect` before rendering a component

ougarte commented on apps/ui/src/components/checkbox/types.ts:
```diff
@@ -0,0 +1,40 @@
+   * 
+   * @default false
+   */
+  indeterminate?: boolean;
```
I recommend using `value?: boolean | null` where `null means intermediate` so that it could be used as

```tsx
<Checkbox />                               // defaults that could be worked as uncontrolled component
<Checkbox value={true} />         // checked
<Checkbox value={false} />.       // unchecked
<Checkbox value={null} />          // intermediate
```

Commit:
Merge branch 'main' of https://github.com/ouspace/gol into feature/GOL-2-checkbox-component
:wq

Commit:
refactor(checkbox): apply suggestions and improve tests with best practices
